smart_card_connector_app/src/application_unittest.cc

```diff
@@ -165,11 +180,117 @@ class SmartCardConnectorApplicationTest : public ::testing::Test {
   }
 
- private:
+  // Sends a simulated JS-to-C++ notification of a PC/SC client being added (in
+  // the real world it's usually another Chrome Extension that wants to access
+  // smart cards).
+  void SimulateJsClientAdded(int handler_id,
+                             const std::string& client_name_for_log) {
+    SimulateFakeJsMessage("pcsc_lite_create_client_handler",
+                          DictValueBuilder()
+                              .Add("handler_id", handler_id)
+                              .Add("client_name_for_log", client_name_for_log)
+                              .Get());
+  }
+
+  // Sends a simulated JS-to-C++ notification of a PC/SC client being removed.
+  void SimulateJsClientRemoved(int handler_id) {
+    SimulateFakeJsMessage(
+        "pcsc_lite_delete_client_handler",
+        DictValueBuilder().Add("handler_id", handler_id).Get());
+  }
+
+  // Sends a simulated JS-to-C++ request to call a PC/SC function.
+  void SimulateCallFromJs(const std::string& requester_name,
+                          RequestId request_id,
+                          const std::string& function_name,
+                          Value arguments) {
+    RemoteCallRequestPayload remote_call_payload;
+    remote_call_payload.function_name = function_name;
+    // Convert an array `Value` to `vector<Value>`. Ideally the conversion
+    // wouldn't be needed, but in tests it's more convenient pass a single
+    // `Value` (e.g., constructed via `ArrayValueBuilder`).
+    remote_call_payload.arguments =
+        ConvertFromValueOrDie<std::vector<Value>>(std::move(arguments));
+
+    RequestMessageData request_data;
+    request_data.request_id = request_id;
+    request_data.payload = ConvertToValueOrDie(std::move(remote_call_payload));
+
+    SimulateFakeJsMessage(GetRequestMessageType(requester_name),
+                          ConvertToValueOrDie(std::move(request_data)));
+  }
+
+  LONG SimulateEstablishContextJsCall(int handler_id,
+                                      RequestId request_id,
+                                      DWORD scope,
+                                      Value reserved1,
+                                      Value reserved2,
+                                      SCARDCONTEXT& out_scard_context) {
+    const std::string requester_name = GetJsClientRequesterName(handler_id);
+    auto waiter =
+        global_context_.CreateResponseWaiter(requester_name, request_id);
+    SimulateCallFromJs(requester_name, request_id,
+                       /*function_name=*/"SCardEstablishContext",
+                       ArrayValueBuilder()
+                           .Add(scope)
+                           .Add(std::move(reserved1))
+                           .Add(std::move(reserved2))
+                           .Get());
+    waiter->Wait();
+    Value reply = std::move(*waiter).take_value();
+
+    GOOGLE_SMART_CARD_CHECK(reply.is_array());
+    GOOGLE_SMART_CARD_CHECK(!reply.GetArray().empty());
```

Comment:
nit: Perhaps here we can put reply.GetArray() into some variable to save some GetArray() calls